src/v/ssx/sharded_ptr.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "base/vassert.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/sharded.hh>
+
+#include <memory>
+
+namespace ssx {
+
+/// A pointer that is safe to share between shards.
+///
+/// The pointer can be reset only by the home shard; other shards shall not
+/// observe the change until the update has run on their reactor.
+///
+/// As such, it is safe to maintain a pointer or reference to the pointee an any
+/// shard until the next yield point.
+template<typename T>
+class sharded_ptr {
+public:
+    sharded_ptr()
+      : _shard{ss::this_shard_id()} {}
+    ~sharded_ptr() {
+        if (*this) {
+            std::default_delete<T>{}(_state.local());
+        }
+    }
+
+    sharded_ptr(sharded_ptr&& other) noexcept = delete;
+    sharded_ptr(sharded_ptr const&) noexcept = delete;
+    sharded_ptr& operator=(sharded_ptr const&) = delete;
+    sharded_ptr& operator=(sharded_ptr&&) = delete;
+
+    T& operator*() const { return *_state.local(); }
+    T* operator->() const { return _state.local(); }
+    explicit operator bool() const {
+        return _state.local_is_initialized() && _state.local() != nullptr;
+    }
+
+    ss::future<> reset(std::unique_ptr<T> u = nullptr) {
+        assert_shard();
+        auto gh{_gate.hold()};
+        auto up{u.get()};
+
+        if (!_state.local_is_initialized()) {
+            if (u == nullptr) {
+                co_return;
+            }
+            co_return co_await _state.start(up).finally(
+              [this, u{std::move(u)}]() mutable noexcept {
+                  if (u.get() == _state.local()) {
```

Comment:
In case `start` fails for some reason, possibly a failed allocation or something like that.

Reply:
Wouldn't you want a similar condition in the finally for other case below (we were don't do the init)?

Overall this might be simpler to reason about if this init path just null-init'd the pointers on all the other shards, and then full through to the update logic below: fewer cases to think about. It's only a one-time inefficiency (per sharded_ptr).

Reply:
Changed: no raw pointers or funky memory management.